### PR TITLE
[Draft] PettingZoo Support

### PR DESCRIPTION
**Pull request by jkterry1**
This is a nonworking draft of adding support for the PettingZoo API for multi-agent RL. This will make it able to be used with a bunch of other multi-agent libraries (SB3 and similar via SuperSuit, the ALL, Tianshou, etc.) RLlib also has native PettingZoo support too. Feel free to take a look and hopefully finish this in the next week.

### CODE REVIEW

**Comment by jkterry1** [comment]
This PR also adds basic CI

**Comment by LucasAlegre** [comment]
Thank you very much for this contribution!
I will take a better look at it this week. The original sumo-rl env works like a ParallelEnv, but I believe it is possible to convert to AECEnv easily.

**Comment by jkterry1** [comment]
That's a ton. I'll be adding support for both APIs by the time I'm done.

Right now this code is partially working, but I've been stuck for a few hours on the current CI error about "libsumo.libsumo.FatalTraCIError: A network was not yet constructed." Could you please take a look into that if you get a chance?

**Comment by LucasAlegre** [comment]
I believe this is happening because there is a call to compute_rewards() in the constructor of SumoEnvironmentPZ. 
But this method only works after a call to reset() has already happened. It is in the reset() method that the simulation starts for real.

**Comment by LucasAlegre** [comment]
@jkterry1 do you need any help with the code?

**Comment by jkterry1** [comment]
@LucasAlegre Thanks a ton. There was an event in my personal life that's been extraordinarily time consuming, so I haven't gotten anything done and probably won't for a little while longer. If you'd be willing to finish hooking up the step function that'd be awesome. Everything else is done or is something I should be able to easily do or is already done.

More pressingly though, would you be willing to different map files do like I mentioned in #44? I don't have enough knowledge of Sumo to figure this out on my own and need it for my research.

**Comment by jkterry1** [comment]
(you should have push access to this PR BTW)

**Comment by LucasAlegre** [comment]
> @LucasAlegre Thanks a ton. There was an event in my personal life that's been extraordinarily time consuming, so I haven't gotten anything done and probably won't for a little while longer. If you'd be willing to finish hooking up the step function that'd be awesome. Everything else is done or is something I should be able to easily do or is already done.
Hope everything is fine! 
Now it is passing the API test :)
But I have a question: is there a way to implement the agent selector in a way that not all agents act every cycle (like asynchronous, every step some agents act and others don't)? In some envs it may happen that a traffic signal can't act because it is waiting the minimum green time.
 
> More pressingly though, would you be willing to different map files do like I mentioned in #44? I don't have enough knowledge of Sumo to figure this out on my own and need it for my research.
I will work on that, I'm not sure if this week, but hopefully soon!

**Comment by jkterry1** [comment]
@LucasAlegre thank you so much for your help with all this! Before this is merged I'd like to have a team mate look at this PR and try basic PettingZoo learning code with it. I should be able to do that Wednesday

**Comment by jkterry1** [comment]
@LucasAlegre btw why is CI failing?

**Comment by jkterry1** [comment]
Also could you please add the relevant wrappers to the environment within pettingzoo for things like proper error message handling and so on like default environments have?

**Comment by LucasAlegre** [comment]
> @LucasAlegre thank you so much for your help with all this! Before this is merged I'd like to have a team mate look at this PR and try basic PettingZoo learning code with it. I should be able to do that Wednesday

No problem! Is there a way to implement the agent selector in a way that not all agents act every cycle (like asynchronous, every step some agents act and others don't)? In some scenarios, it may happen that a traffic signal can't act because it is waiting the minimum green time. Perhaps this can be a problem in the current implementation.

**Comment by LucasAlegre** [comment]
> @LucasAlegre btw why is CI failing?

pytest was not finding any function with "test" in its name. Fixed now.

**Comment by LucasAlegre** [comment]
> Also could you please add the relevant pettingzoo wrappers for things like proper error message handling and so on like built in environments have?

Done!

**Comment by jkterry1** [comment]
Sorry for the delayed reply. Regarding the agent selector, variable numbers of agents (and especially agents that pop in and out of life) is really problematic for most learning code, and my preference is for the agent to take an action that does nothing by default (in fact for my purposes it has to work that way). That being said, our API does support it (you just remove it from the agents list, look at how death is handled in the knights archers zombies environment in PettingZoo for reference). If you do, could you make it an environment argument (that defaults to off)?

**Comment by LucasAlegre** [comment]
> Sorry for the delayed reply. Regarding the agent selector, variable numbers of agents (and especially agents that pop in and out of life) is really problematic for most learning code, and my preference is for the agent to take an action that does nothing by default (in fact for my purposes it has to work that way). That being said, our API does support it (you just remove it from the agents list, look at how death is handled in the knights archers zombies environment in PettingZoo for reference). If you do, could you make it an environment argument (that defaults to off)?

In the current implementation, if it is not the time for the agent to act yet, the action will do nothing and it will return the previous observation and reward again. Is this the behaviour you expect?

**Comment by jkterry1** [comment]
The behavior I'd expect is that there is no reward no matter what and the observation is whatever is current.

**Comment by LucasAlegre** [comment]
Hey, with my new commit I added the file 'experiments/ql_4x4grid_pz.py' which reproduces 'experiments/ql_4x4grid.py' but with PettingZoo env.
If you are ok with it, I can merge :)

**Comment by jkterry1** [comment]
@LucasAlegre I think this is good; are you able to change the reward an observation behavior how I described above?

File: sumo_rl/environment/env.py [hunk 1]
```diff
@@ -207,7 +246,88 @@ def _discretize_elapsed_time(self, elapsed):
         for i in range(self.max_green//self.delta_time):
             if elapsed <= self.delta_time + i*self.delta_time:
                 return i
-        return self.max_green//self.delta_time -1
+        return self.max_green//self.delta_time - 1
 
 
-    
+class SumoEnvironmentPZ(AECEnv, EzPickle):
+    metadata = {'render.modes': [], 'name': "sumo_rl_v0"}
```
Comment (benblack769) [inline sumo_rl/environment/env.py:253]: 'human' should be in the render modes list

File: sumo_rl/environment/env.py [hunk 2]
```diff
@@ -207,7 +246,88 @@ def _discretize_elapsed_time(self, elapsed):
...
+
+    def reset(self):
+        self.env.reset()
+        self.agents = self.possible_agents[:]
+        self.agent_selection = self._agent_selector.reset()
+        self.rewards = {agent: 0 for agent in self.agents}
+        self._cumulative_rewards = {agent: 0 for agent in self.agents}
+        self.dones = {agent: False for agent in self.agents}
+        self.infos = {agent: {} for agent in self.agents}
+        return self.env.observations.copy()
```
Comment (benblack769) [inline sumo_rl/environment/env.py:286]: reset should not return anything

File: sumo_rl/environment/env.py (hunk 1 above)
Comment (benblack769) [reply sumo_rl/environment/env.py:253]: Unless the base environment truely has no rendering support I suppose.

File: sumo_rl/environment/env.py [hunk 3]
```diff
@@ -207,7 +246,88 @@ def _discretize_elapsed_time(self, elapsed):
...
+            self.env._run_steps()
+            self.env._compute_observations()
+            new_rewards = self.env._compute_rewards()
+            self.rewards.update(new_rewards)
+        else:
+            self._clear_rewards()
+        
+        self.dones = self.env._compute_dones()
+        if self.dones['__all__']:
+            self.agents = ()
```
Comment (benblack769) [inline sumo_rl/environment/env.py:321]: I feel like this logic shouldn't work. The agents should be removed one at a time with the `_was_done_step` helper method.

Comment (benblack769) [reply sumo_rl/environment/env.py:321]: I.e., there should be no special code that is needed to remove agents.

**Comment by jkterry1** [comment]
@LucasAlegre

Also, in the current structure, the pettingzoo class inherits from the RLlib env class. This is not ideal because it means that people who are training on these environments with libraries other than RLlib (like me) still have it install it, and RLlib is an incredibly onerous dependency. Given that RLlib has full pettingzoo support (as do a lot of other MARL libraries), how would you feel about only having a pettingzoo environment and just changing the examples to use the pettingzoo class instead of having a pettingzoo specific example?

File: sumo_rl/environment/env.py [hunk 4]
```diff
@@ -207,7 +246,88 @@ def _discretize_elapsed_time(self, elapsed):
...
+        del self.dones['__all__']
+
+        self.agent_selection = self._agent_selector.next()
+        self._cumulative_rewards[agent] = 0
+        self._accumulate_rewards()
+
+def make_env(**kwargs):
+    env = SumoEnvironmentPZ(**kwargs)
+    env = wrappers.CaptureStdoutWrapper(env)
```
Comment (benblack769) [inline sumo_rl/environment/env.py:330]: This wrapper allows results that are printed to the terminal to be returned as a string. Is that the intended use case here?

File: tests/pz_test.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+from sumo_rl import SumoEnvironmentPZ, make_env
+
+
+def test_api():
+    env = make_env(net_file='nets/4x4-Lucas/4x4.net.xml',
+                   route_file='nets/4x4-Lucas/4x4c1c2c1c2.rou.xml',
+                   out_csv_name='outputs/4x4grid/a3c',
+                   use_gui=False,
+                   num_seconds=80000)
+    api_test(env)
```
Comment (benblack769) [inline tests/pz_test.py:11]: Perhaps a render test and seed test should be added here if appropriate?

Comment (jkterry1) [reply tests/pz_test.py:11]: probably, yes

File: sumo_rl/environment/env.py (hunk 3 above)
Comment (benblack769) [reply sumo_rl/environment/env.py:321]: I think the correct logic here is 
```
if self.dones['__all__']:
    del self.dones['__all__']
    for a in self.dones:
         self.dones[a] = True
```

File: sumo_rl/environment/env.py (hunk 1 above)
Comment (LucasAlegre) [reply sumo_rl/environment/env.py:253]: To use SUMO-GUI you have to initialize the env with use_gui=True.
Unfortunately, there is no way to make the rendering and the env logic independent with SUMO. 
So render() do nothing. Should I add 'human' to metadata anyway?

File: sumo_rl/environment/env.py (hunk 2 above)
Comment (LucasAlegre) [reply sumo_rl/environment/env.py]: In 'https://github.com/PettingZoo-Team/PettingZoo/blob/master/pettingzoo/utils/env.py':
    def reset(self):
        '''
        resets the environment and returns a dictionary of observations (keyed by the agent name)
        '''
        raise NotImplementedError
     This doc should be changed.

Comment (LucasAlegre) [reply sumo_rl/environment/env.py]: Anyway, I removed it in https://github.com/LucasAlegre/sumo-rl/pull/45/commits/a39fdde0773eb8f5ed66796dca3531c6ce88e37b

File: sumo_rl/environment/env.py (hunk 3 above)
Comment (LucasAlegre) [reply sumo_rl/environment/env.py]: Thanks! Fixed it in https://github.com/LucasAlegre/sumo-rl/pull/45/commits/9f91bec2c8806722252c864d96e43c83fe266c4a

File: sumo_rl/environment/env.py (hunk 4 above)
Comment (LucasAlegre) [reply sumo_rl/environment/env.py]: I guess this was unnecessary, thanks! Removed it in https://github.com/LucasAlegre/sumo-rl/pull/45/commits/1cd9f978b0af2b96e22d99e535582d15fc8722cb

**Comment by LucasAlegre** [comment]
> @LucasAlegre
> 
> Also, in the current structure, the pettingzoo class inherits from the RLlib env class. This is not ideal because it means that people who are training on these environments with libraries other than RLlib (like me) still have it install it, and RLlib is an incredibly onerous dependency. Given that RLlib has full pettingzoo support (as do a lot of other MARL libraries), how would you feel about only having a pettingzoo environment and just changing the examples to use the pettingzoo class instead of having a pettingzoo specific example?

That is something I was already thinking about!
In my last commit https://github.com/LucasAlegre/sumo-rl/pull/45/commits/1cbc8ea42af1e0b0a15ef934fb4d34f5fcf880aa I removed the rllib dependence. I also made the agents act synchronously, as I think it is better this way.

**Comment by LucasAlegre** [comment]
@jkterry1 pytest is failing because of:
```
/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pettingzoo/utils/save_observation.py:5: in <module>
    from PIL import Image
E   ModuleNotFoundError: No module named 'PIL'
```
I guess you should add PIL as pettinzoo dependence.

**Comment by jkterry1** [comment]
@LucasAlegre 

Thanks a ton! A few things:

-The PIL issue will be fixed in the next pettingzoo release (in a few weeks), can you please just add it as a dependency for now so tests properly function?
-Did you see my comment about the expected behavior of rewards and observations above?
-What do you mean by "I also made the agents act synchronously, as I think it is better this way."?

**Comment by LucasAlegre** [comment]
> @LucasAlegre
> 
> Thanks a ton! A few things:
> 
> -The PIL issue will be fixed in the next pettingzoo release (in a few weeks), can you please just add it as a dependency for now so tests properly function?

Done!
> -Did you see my comment about the expected behavior of rewards and observations above?
> -What do you mean by "I also made the agents act synchronously, as I think it is better this way."?

I made all traffic signal agents act synchronously every 'delta' sumo time-steps. Then, if an agent can't change phase because it has not passed 'min_green' seconds yet, it will just keep the same phase no matter the action is. It solves the problem of agents popping in and out of life, and thus addresses your comment about the expected behavior of rewards and observations.

**Comment by jkterry1** [comment]
That sounds great, but just to confirm, what's the observation and reward for a light that can't change phase?

I'll run a last round of tests in the morning.

**Comment by LucasAlegre** [comment]
> That sounds great, but just to confirm, what's the observation and reward for a light that can't change phase?
> 
> I'll run a last round of tests in the morning.

The observation is whatever it is at the moment, and the reward too. You can think of it as a state where all actions have the same effect. It does not affect learning.

**Comment by jkterry1** [comment]
I don't believe that giving whatever reward the agent can get at the moment, when it was not able to act, is an expected behavior? This generally is not how reward works in reinforcement learning.

**Comment by LucasAlegre** [comment]
> I don't believe that giving whatever reward the agent can get at the moment, when it was not able to act, is an expected behavior? This generally is not how reward works in reinforcement learning.

Sorry if I was not clear. It is not giving whatever reward the agent can get at the moment. It receives the actual reward as a consequence of its action.
Example: Suppose there are two actions/phases, A1 and A2. A1 is currently active, but it has not already passed min_green seconds yet. Now, if the agent selects A1 or A2, A1 will stay active anyway and the reward will be the difference between the delay before and after the action was executed (as in the Readme).

An analogy of this situation is a grid world where the agent is facing an upper-right corner wall. If he moves UP or RIGHT, it will have the same effect and it will receive the reward of the current cell.

**Comment by LucasAlegre** [comment]
> Hey, so I just sat down to run tests and I added the specific learning file that I'll be using with "sb3.py" in experiments. I had a few questions and issues that came up in the process:
> 
> -I know you said that you'd document the different environment xml files you created eventually, but in the mean time how many agents are in the 4x4 environment I'm using for testing so I can properly space evaluations?

There are 16 agents in the 4x4 grid.

> -How many steps are you training these environments for, just so I have reference? Maybe I'm blind but I'm not seeing it defined in your rllib learning files.

100 k steps should be enough.

> -Looking at the code you've added, it seems like the way to create PZ environments is by using the "from sumo_rl import make_env" and using the make_env method and the way to create RLlib API environments is with "from sumo_rl import SumoEnvironment" function and using the "SumoEnvironment" method. This is kind of confusing behavior to me? Additionally, if that's the case there are unused "SumoEnvironmentPZ" imports

To use RLlib API I'm using the RLlib wrapper, check the file experiments/a3c_4x4grid.py in this PR. But I agree, I will update the imports.

> -Some changes probably need to be made to the readme need to be made in light of this PR, e.g. you have the line "The main class SumoEnvironment inherits MultiAgentEnv from RLlib"

I already did, now it says "The main class SumoEnvironment behaves like a MultiAgentEnv from RLlib." But I will improve the README when I add the new environments.

> -My test learning code fails, seemingly due to an error with how you added the pettingzoo internal wrappers with this error:
> 
> ```
> jkterry@prophet:~/sumo-rl$ python3 experiments/sb3.py
> /home/jkterry/.local/lib/python3.6/site-packages/supersuit/__init__.py:20: UserWarning: You're using SuperSuit 3.0, released 7/7/21. The entire codebase has been rewritten or refactored as part of this release. While we've tested it thoroughly, please ensure everything you're doing still works properly and report any issues at https://github.com/PettingZoo-Team/SuperSuit. This warning will be removed 2 months after release.
>   warnings.warn("You're using SuperSuit 3.0, released 7/7/21. The entire codebase has been rewritten or refactored as part of this release. While we've tested it thoroughly, please ensure everything you're doing still works properly and report any issues at https://github.com/PettingZoo-Team/SuperSuit. This warning will be removed 2 months after release.")
> Traceback (most recent call last):
>   File "experiments/sb3.py", line 24, in <module>
>     env = env.parallel_env()
>   File "/home/jkterry/.local/lib/python3.6/site-packages/pettingzoo/utils/wrappers/order_enforcing.py", line 45, in __getattr__
>     f"'{type(self).__name__}' object has no attribute '{value}'"
> AttributeError: 'OrderEnforcingWrapper' object has no attribute 'parallel_env'
> ```

This is because the method .parallel_env() does not exist. Where should it come from?

**Comment by LucasAlegre** [comment]
@jkterry1 I also notice that RLLib PettingZooEnv wrapper does not work together with the OrderEnforcingWrapper:
```
File "/home/lucas/miniconda3/lib/python3.7/site-packages/ray/rllib/env/pettingzoo_env.py", line 72, in __init__
    self.agents = self.aec_env.agents
  File "/home/lucas/miniconda3/lib/python3.7/site-packages/pettingzoo/utils/wrappers/order_enforcing.py", line 42, in __getattr__
    raise AttributeError(f"{value} cannot be accessed before reset")
AttributeError: agents cannot be accessed before reset
```

**Comment by benblack769** [comment]
So the parallel_env() issue is that in pettingzoo, there are these modules with a env() and parallel_env() functions. https://github.com/PettingZoo-Team/PettingZoo/blob/master/test/example_envs/generated_agents_parallel_v0.py#L12 Environments don't spawn parallel environments, modules do. This isn't a requirement of the API, it is just a convention. If you want to turn an  existing environment into a parallel environment, then you can just use
```
from pettingzoo.utils.conversions import to_parallel
to_parallel(env)
```

**Comment by benblack769** [comment]
As for the rllib issue, what version of rllib are you using? That appears to be a fairly old version.

**Comment by jkterry1** [comment]
@LucasAlegre Could you please add the parallel extension that ben described with the other changes?

**Comment by LucasAlegre** [comment]
> As for the rllib issue, what version of rllib are you using? That appears to be a fairly old version.

Updating solved the issue, thanks!

**Comment by LucasAlegre** [comment]
@jkterry1 @benblack769 
The code in 'experiments/sb3.py' is running now, but there are the following "issues":
* When using LIBSUMO, it is not possible to instantiate more than 1 simulation at the same time. This means that eval_callback can't work. If you use TRACI, I can easily implement multi-client support and this would be possible. But remember that TRACI is way slower than LIBSUMO, so I'm not sure whether this is advantageous.
* As I explained before, render() does nothing. You need to instantiate the env with 'use_gui=True' to run the SUMO-GUI and watch the simulation.

I believe we could merge this PR, as there are already many changes. Next I can:
* Add the new environments with their documentations (that should be really quick, I already have the sumo files)
* Improve README and examples.
* Implement the multi-client support in case you really need it. (this should be done changing very few lines of code)

What do you think?

**Comment by jkterry1** [comment]
If you want to merge now and open a new PR that's fine with me, I'll reply tonight about the render and environment duplication problems.

**Pull request merged**